### PR TITLE
Refactoring after implementing the verify commits hook

### PR DESCRIPTION
**Pull request by JWorthe-jemstep**

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -46,34 +46,29 @@ pub struct PrePush {
...
-
-pub fn prepare_commit_msg<F: Fs, G: Git>(opt: PrepareCommitMsg, config: Config) -> Result<(), Box<dyn Error>> {
+pub fn prepare_commit_msg<F: Fs, G: Git>(opt: PrepareCommitMsg, config: Config) -> Result<PolicyResult, Box<dyn Error>> {
     if opt.commit_source.is_none() {
-        if let Some(_) = config.prepend_branch_name {
-            prepend_branch_name::<F, G>(opt.commit_file)?;
-        }
-
-        Ok(())
+        compose(vec![
```
Comment (jemstep-edward) [inline src/lib.rs]: Maybe call this collect?

File: src/main.rs [hunk 2]
```diff
@@ -47,64 +49,100 @@ enum Command {
...
+        .expect("ERROR: Logger was initialized twice");
     
     info!("{}", block("Ahoy, maties! Welcome to Capn Githook!"));
 
-    info!("Read configuration file");
-    let config = match LiveGit::new()?.read_config() {
+    let git = match LiveGit::new() {
+        Ok(g) => g,
+        Err(e) => {
+            error!("Capn Githook must be called in a Git repo. Error: {}", e);
```
Comment (jemstep-edward) [inline src/main.rs]: Log the error first and then make the suggestion that it should be called from inside of a git repo.

File: src/main.rs [hunk 3]
```diff
@@ -47,64 +49,100 @@ enum Command {
...
+        Err(e) => {
+            error!("System error: {}", e);
+            exit(1);
+        }
+    }
+}
+
+fn execute_command(command: Command, config: Config) -> Result<PolicyResult, Box<dyn Error>> {
+    match command {
+        Command::PrepareCommitMsg(x) => {
```
Comment (jemstep-edward) [inline src/main.rs]: What is x??

File: src/main.rs [hunk 4]
```diff
@@ -47,64 +49,100 @@ enum Command {
...
+            info!("Calling prepare-commit-msg");
+            prepare_commit_msg::<LiveFs, LiveGit>(x, config)
+        },
         Command::PrePush(x) => {
```
Comment (jemstep-edward) [inline src/main.rs]: ditto

File: src/policies.rs [hunk 5]
```diff
@@ -80,75 +127,73 @@ pub fn verify_git_commits<G: Git, P: Gpg>(config: &VerifyGitCommitsConfig, old_v
...
+                    },
+                    Err(e) => Err(e)
                 }
             }
-        }
-    }
-    Ok(())
+        })
+        .collect::<Result<Vec<PolicyResult>, Box<dyn Error>>>()
+        .map(|xs| xs.into_iter().fold(PolicyResult::Ok, |acc, x| acc.and(x)))
```
Comment (jemstep-edward) [inline src/policies.rs]: Maybe try just writing `and` instead.

File: src/policies.rs [hunk 6]
```diff
@@ -80,75 +127,73 @@ pub fn verify_git_commits<G: Git, P: Gpg>(config: &VerifyGitCommitsConfig, old_v
...
+            debug!("Verify author, committer email addresses for commit {}", commit.id());
+            match (commit.author().email(), commit.committer().email()) {
+                (None, _) => PolicyResult::MissingAuthorEmail(commit.id()),
+                (_, None) => PolicyResult::MissingCommitterEmail(commit.id()),
+                (Some(s), _) if !s.ends_with(&format!("@{}", author_domain)) => PolicyResult::InvalidAuthorEmail(commit.id(), s.to_string()),
+                (_, Some(s)) if !s.ends_with(&format!("@{}", committer_domain)) => PolicyResult::InvalidCommitterEmail(commit.id(), s.to_string()),
+                _ => PolicyResult::Ok
+            }
+        })
+        .fold(PolicyResult::Ok, |acc, x| acc.and(x))
```
Comment (jemstep-edward) [inline src/policies.rs]: ditto

File: src/lib.rs (hunk 1 above)
Comment (JWorthe-jemstep) [reply src/lib.rs]: This triggered the idea to implement FromIterator for PolicyResult

**Pull request merged**